driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -7,177 +7,144 @@ use crate::{
+    ) -> Result<()> {
+        log::info!("Received {} events.", events.len());
+        let block_hashes = events
+            .iter()
+            .map(|event| {
+                let metadata = event
+                    .meta
+                    .as_ref()
+                    .ok_or_else(|| anyhow!("event without metadata: {:?}", event))?;
+                Ok(metadata.block_hash)
+            })
+            .collect::<Result<HashSet<H256>>>()?;
+        behind_mutex
+            .block_timestamp_reader
+            .prepare_cache(block_hashes)
+            .await?;
+        behind_mutex
+            .orderbook
+            .delete_events_starting_at_block(delete_events_starting_at_block);
+        for event in events {
+            self.handle_event(behind_mutex, event).await?;
+        }
+        log::info!("Finished applying events");
+        Ok(())
     }
-    log::info!("Finished applying events");
-    Ok(())
-}
 
-/// Apply a single event to the orderbook.
-async fn handle_event(
-    orderbook: &mut Orderbook,
-    block_timestamp_reader: &mut impl BlockTimestampReading,
-    event: Event<batch_exchange::Event>,
-) -> Result<()> {
-    match event {
-        Event {
-            data,
-            meta: Some(meta),
-        } => {
-            let block_timestamp = block_timestamp_reader
-                .block_timestamp(meta.block_hash)
-                .await?;
-            orderbook.handle_event_data(
+    /// Apply a single event to the orderbook.
+    async fn handle_event(
+        &self,
+        behind_mutex: &mut BehindMutex,
+        event: Event<batch_exchange::Event>,
+    ) -> Result<()> {
+        match event {
+            Event {
                 data,
-                meta.block_number,
-                meta.log_index,
-                meta.block_hash,
-                block_timestamp,
-            );
+                meta: Some(meta),
+            } => {
+                let block_timestamp = behind_mutex
+                    .block_timestamp_reader
+                    .block_timestamp(meta.block_hash)
+                    .await?;
+                behind_mutex.orderbook.handle_event_data(
+                    data,
+                    meta.block_number,
+                    meta.log_index,
+                    meta.block_hash,
+                    block_timestamp,
+                );
+            }
+            Event { meta: None, .. } => bail!("event without metadata"),
         }
-        Event { meta: None, .. } => bail!("event without metadata"),
+        Ok(())
+    }
+}
+
+impl StableXOrderBookReading for UpdatingOrderbook {
+    /// Blocks on updating the orderbook.
+    fn get_auction_data(&self, batch_id_to_solve: U256) -> Result<(AccountState, Vec<Order>)> {
+        let mut behind_mutex = self
+            .behind_mutex
+            .lock()
+            .map_err(|err| anyhow!("poison error: {}", err))?;
+        futures::executor::block_on(self.update_with_events(&mut behind_mutex))?;
```

Comment:
Would it make sense to make `update_with_events` a synchronous method instead? Would we then still need the mutex?

Reply:
We always need the mutex because all orderbooks must be `Sync`. Instead of using the mutex here we could wrap the whole thing with a Mutex which would make this implementation slightly simpler.

Reply:
I might not understand the full implications of this, but my suggestion would be to synchronously wait on the "potentially long running" futures using our [util extension](https://github.com/gnosis/dex-services/blob/affcdbb27d20f1b64c0d01dc2d15af235311ec9a/driver/src/util.rs#L44-L46).

I see how `UpdatingOrderbook` itself must be Sync, but I don't understand why we need a Mutex for this (couldn't we use an arc option or the like)? Mutexes are a pretty heavy tool (opening the potential for deadlocks) which I'd like to avoid if not really necessary.

Reply:
There *has* to be a mutex because `get_auction_data` can be called from multiple threads. Arc only guarantees that the data inside lives long enough but not that accessing it is safe.
The worry about deadlocks is misplaced because we can see that the mutex is only locked in one place.

Reply:
I see now how we need the Mutex here in order to have interior mutability and Sync at the same time.

> The worry about deadlocks is misplaced because we can see that the mutex is only locked in one place.

Theoretically we could have a re-entrancy that causes the deadlock. In general, reasoning about concurrent code is extremely hard and commonly underestimated (it's easy for the initial author, but can easily evolve into a nightmare). That's why, if possible I'd like to avoid it.

Reply:
Correct me if I'm wrong, but a `Mutex` is also needed for interior mutability. Since we are sharing the orderbook across threads which potentially mutate (by adding new events for example) the orderbook, a `Mutex` is indeed needed. Just an `Arc` would allow you to access a reference to the underlying data but not a mutable reference which we need here.

Reply:
> Theoretically we could have a re-entrancy that causes the deadlock. In general, reasoning about concurrent code is extremely hard 

This is a very true statement. Concurrent code is always hard especially when you didn't write it.

One solution to this is to have the orderbook updating running on its own thread and use message passing to perform updates and retrieve the current orderbook. That way all orderbook updates happen in one thread and "get orderbook" messages are sent over a channel to that thread.

Reply:
That being said, I am happy with the Mutex use here, it is fairly straightforward and its use isn't spread over multiple files.

Reply:
Agreed, I didn't immediately see the need for interior mutability and given this it seems like there is no real simple better option (I'd like to avoid more scattered threading).